README
TODO: Write readme

= NOTE
Evaluation of configuration MAX_COLLECT_SIZE=$((2*1024**2)) only works with /bin/sh linked to bash (not dash!).


debian/changelog
isibackup (2.0.36) all; urgency=low

  * New release
  * Enabled postgresql backup and changed condition for creating mysql
    and postgresqlbackup (based on installed package and not on existing
    directory).

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Mon, 18 Apr 2011 14:00:04 +0200

isibackup (2.0.35) all; urgency=low

  * New release
  * Removed unwanted exit 0

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Fri, 14 Jan 2011 14:36:08 +0100

isibackup (2.0.34) all; urgency=low

  * New release
  * Added debug logs and fixed mount_points mechanism

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Fri, 14 Jan 2011 14:19:47 +0100

isibackup (2.0.33) all; urgency=low

  * New release
  * test
  * Added notice for using bash not dash
  * Removed backup file
  * first commit

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Thu, 16 Dec 2010 19:14:28 +0100

isibackup (2.0.32) unstable; urgency=low

  * New release

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Thu, 18 Nov 2010 16:13:22 +0100

isibackup (2.0.31) all; urgency=low

  * New release
  * Removed some more dependencies

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Fri, 13 Aug 2010 15:38:40 +0200

isibackup (2.0.30) all; urgency=low

  * New release
  * Removed some dependencies for running on lucid

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Fri, 13 Aug 2010 15:33:55 +0200

isibackup (2.0.29) all; urgency=low

  * New release
  * Exit on unexpected pathlist entry
  * readlink -f for OPT_TARGET_DIR and OPT_SOURCE_DIR
  * Fixed variable name when "OPT_RECREATE_ARCHIVE" is true.
  * Do not exit if pack compress encrypt command fails, report error
    only.

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Wed, 14 Apr 2010 16:35:19 +0200

isibackup (2.0.28) all; urgency=low

  * New release
  * Fixed dependencies to libisi

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Thu, 14 Jan 2010 14:41:49 +0100

isibackup (2.0.27) unstable; urgency=low

  * New release
  * Changed media glob pattern to work with dash and bash.

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Tue, 05 Jan 2010 14:19:21 +0100

isibackup (2.0.26) all; urgency=low

  * New release
  * Corrected set path not to point to the link.
  * Exclude pbuilder cache directory from system backup
  * Exclude googleearcht cache from data backup

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Mon, 28 Dec 2009 11:05:31 +0100

isibackup (2.0.25) all; urgency=low

  * New release
  * Added --silent deprecated warning.
  * Isirestore mit tar und files anderer packager werden
    aurger[_\c3_][_\a4_]umt.
  * First version with tar packer.
  * Try if /var/log/isibackup is writable.
  * Made tar working.
  * Skip and warn if cpio should handle 4G Files.

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Tue, 10 Nov 2009 10:48:42 +0100

isibackup (2.0.24) all; urgency=low

  * New release
  * Added logs to /var/log/isibackup

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Wed, 28 Oct 2009 10:59:24 +0100

isibackup (2.0.23) unstable; urgency=low

  * New release
  * Removed dependency from log4r
  * Simplified isibackup script (removed secundary logging).

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Wed, 28 Oct 2009 10:49:01 +0100

isibackup (2.0.22) etch; urgency=low

  * New release
  * Precreate log file to avoid unfinished writing on full disks

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Mon, 05 Oct 2009 10:49:03 +0200

isibackup (2.0.21) etch; urgency=low

  * New release
  * Give default sets if no sets can be found by backup.rb
  * Added missingok to logrotate.

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Wed, 26 Aug 2009 16:57:04 +0200

isibackup (2.0.20) etch; urgency=low

  * New release
  * Do ignore state lines that do not match exact count of fields.

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Wed, 26 Aug 2009 10:43:48 +0200

isibackup (2.0.19) etch; urgency=low

  * New release
  * Fixed incremental mysql backup.

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Tue, 09 Jun 2009 22:47:45 +0200

isibackup (2.0.18) etch; urgency=low

  * New release
  * Isirestore for directories.
  * Fixed copy bug.

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Thu, 28 May 2009 10:06:45 +0200

isibackup (2.0.17) unstable; urgency=low

  * New release
  * Small fix for :copy mode.

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Mon, 25 May 2009 10:10:17 +0200

isibackup (2.0.16) unstable; urgency=low

  * New release
  * Reenabled copy mode with collect code.

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Tue, 19 May 2009 10:25:11 +0200

isibackup (2.0.15) etch; urgency=low

  * New release

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Mon, 18 May 2009 18:32:31 +0200

isibackup (2.0.14) unstable; urgency=low

  * Added to_s function to backup.rb
  * Ignore backups in isirestore with have missing incrementals.
  * Tidied up changlog after isirelease added all changes.

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Mon, 18 May 2009 18:32:03 +0200

isibackup (2.0.12) etch; urgency=low

  * work on isibackup-dialog

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Wed, 06 May 2009 10:35:15 +0200

isibackup (2.0.11) etch; urgency=low

  * Claim errors when isibackup-main did not successfuly exit in isibackup. (exitvalue)
  * Added isibackup-dialog script

 -- Flavio Pellanda (Logintas 2008) <flavio.pellanda@logintas.ch>  Tue, 05 May 2009 16:26:38 +0200

isibackup (2.0.9) etch; urgency=low

  * fixed bug that control file could not be parsed


lib/backup.rb
require "pathname"
require "path_list"
class Backup
  def Backup.root
    (ENV["BACKUP_ROOT"] or (defined?(BACKUP_ROOT) and BACKUP_ROOT))
  end

  def Backup.update_options(options)
    options = options.dup
    options[:backup_root] ||= Backup.root
    options[:host_name] ||= (ENV["HOST"] or (defined?(HOST) and HOST))
    options[:domain] ||= (ENV["NET"] or (defined?(NET) and NET))

    raise "Backup root not defined." if options[:backup_root].nil? or options[:backup_root] == ""
    options[:backup_root] = Pathname.new(options[:backup_root]).cleanpath
    raise "Backup root not found '#{options[:backup_root]}'" if !options[:backup_root].exist?

    options[:state_path] ||= options[:backup_root] + "state"
    options[:state_path] = Pathname.new(options[:state_path])
#    raise "State path not found '#{options[:state_path]}'" if !options[:state_path].exist?

    options.keys.each {|k|
      raise "#{k} not defined" if options[k].nil? or options[k] == ""
    }
    options
  end

  def Backup.glob_directories(hosts = nil, modes = nil, sets = nil, options = {})
    options = update_options(options)
    h = {}
    if hosts.nil? or hosts.length == 0
      pattern = "*" 
    else 
      pattern = "{" + hosts.join(",") + "}"
    end

    if sets.nil? or hosts.length == 0
      pattern = "*" 
    else 
    end
    
    sets = Backup.sets if sets.nil?
    sets = "{" + sets.join(",") + "}"
    modes = ["full","incr","diff"] if modes.nil?
    
    $log.debug("Looking for directories for hosts '#{pattern}' and modes #{modes}.")
    dirs = []
    modes.select {|m| m == "full"}.each {|mode|
      glob = "#{options[:backup_root]}/#{sets}/#{mode}/*/#{pattern}"
      $log.debug("Globbing '#{glob}'.")
      Dir.glob(glob).each {|d|
	if d =~ /^.*\/([^\/]+)\/(full)\/([^\/]+)\/([^\/]+)$/
	  dirs.push({:dir => d, :set => $1, :mode => $2, :domain => $3, :date => nil, :host_name => $4})
	else
	  #        dirs.push([:dir => d])
	  $log.warn("Illegal directory found #{d}")
	end
      }
    }
    modes.select {|m| m == "incr" or m == "diff"}.each {|mode|
      glob = "#{options[:backup_root]}/#{sets}/#{mode}/*/*/#{pattern}"
      $log.debug("Globbing '#{glob}'.")
      Dir.glob(glob).each{|d|
	if d =~ /^.*\/([^\/]+)\/(incr|diff)\/([^\/]+)\/(\d\d\d\d-\d\d-\d\d)\/([^\/]+)$/
	  dirs.push({:dir => d,  :set => $1, :mode => $2, :domain => $3, :date => $4, :host_name => $5})
	else
	  $log.warn("Illegal directory found #{d}")
	  #        dirs.push([:dir => d])
	end
      }
    }
    dirs
  end

  def Backup.sets(options = {})
    options = update_options(options)
    root = options[:backup_root]
    glob_path = options[:backup_root] + "*/full/#{options[:domain]}/#{options[:host_name]}"
    $log.debug{"Globbing for sets with: #{glob_path}"}
    sets = Dir.glob(glob_path).map {|dir|
      dir.gsub(options[:backup_root].to_s,"").split("/")[1]
    }
    sets = ["data","config","system"] if sets.length == 0
    $log.debug("Found the following sets: #{sets.inspect}")
    sets
  end

  # config
  def Backup.load_defaults 
    # Default configuration
    set_config("ISIBACKUP_CONFIGDIR",get_config("ISIBACKUP_CONFIGDIR","/etc/isibackup"))
    set_config("ISIBACKUP_CONFIG_DEFAULTS",get_config("ISIBACKUP_CONFIG_DEFAULTS","#{get_config("ISIBACKUP_CONFIGDIR")}/defaults.conf"))
    
    source("#{get_config("ISIBACKUP_CONFIG_DEFAULTS")}")
  end
  # Load configuration
  def Backup.load_config 
  # Config file option:
    # 1. command line
    # 2. default: isibackup.conf

    # Source config, abort if config file cannot be sourced
    conffile = get_config("ISIBACKUP_CONFIG")
    $log.info{"Reading configuration '#{conffile}'"}
    source(conffile)
  end

  def Backup.set_config(name, value)
    raise "Configuration name is nil." if name.nil?
    raise "Cannot set something else than a string to #{name}: #{value.class.name}" if
      !value.nil? and value.class.name != "String"
    ENV[name] = value
    $log.config("ENV['#{name}'] is now: #{ENV[name].inspect}") if $log.respond_to?(:config)
    value
  end
  def Backup.get_config(name, default = nil)
    raise "Configuration name is nil." if name.nil?
    raise "Name must be a string: #{name.inspect}" unless name.class.name == "String"
    if ENV[name].nil? or ENV[name] == ""
      $log.config{"Using default config value #{default.inspect} for #{name}."} if $log.respond_to?(:config)
      return default
    end
    $log.config{"Configuration #{name} has value #{ENV[name].inspect}"} if $log.respond_to?(:config)
    ENV[name]
  end

  # Backup control configuration is 
  # stored here:
  CONTROL_FILE="/etc/isibackup/control" unless defined?(CONTROL_FILE)

  def Backup.control
    configs = {}
    open(CONTROL_FILE,"r") {|f|
      f.each {|line|
	case line
	when /^\s*\#/, /^\s*$/
	  # ignore
	when /^\s*(\S+)\s*\:\s*(\d+|\*|\*\/\d+)\s*\,\s*(\d+|\*|\*\/\d+)\s*\,\s*(\d+|\*|\*\/\d+)\s*$/
	  raise "Multiple definitions for confiuration '#{$1}' found in '#{CONTROL_FILE}'" if configs[$1]
	  configs[$1.to_sym] = {
	    :full => parse_interval($2),
	    :incr => parse_interval($3),
	    :diff => parse_interval($4)
	  }
	else
	  raise "Unexpected line in '#{CONTROL_FILE}': #{line.inspect}"
	end     
      } 
    }
    if (media = configs.delete(:media))      
      if medias = get_config("MEDIAS")
	medias = Dir.glob(medias)
	medias.each {|m| path = Pathname.new(m)
	  if path.mountpoint?
	    configs["media-#{path.basename.to_s}".to_sym] = media.dup
	  end
	}
      end
    end
    
    configs
  end


  def initialize(set, options = {})
    raise "Set is nil." if set.nil?

    options = Backup.update_options(options)
    @host = options[:host_name]
    @domain = options[:domain]
    @set = set
    @root = Pathname.new(options[:backup_root])
    raise "Backup root does not exist '#{@root}'" unless @root.directory?
    @state_path = options[:state_path]
    @options = options
  end

  def to_s
    "Backup of #{@host} #{@set}"
  end
  
  def stamp_file(mode)
    @state_path + "#{@host}.#{@domain}-#{@set}-#{mode}.date"
  end
  def state_log
    @state_path + "#{@host}.#{@domain}.log"
  end
  
  def get_stamp(mode, options = {})
    case mode
    when "full"
      return nil    
    when "diff"
      lStampFName = stamp_file("full")
    when "incr"
      lStampFName_incr=stamp_file("incr")
      lStampFName_full=stamp_file("full")

      #use the timestamp from a previous backup if the last incremental backup was today and not using dirnumbering
      if !@options[:dir_numbering] and
	  !@options[:dir_with_time] and
	  lStampFName_incr.exist? and
	  lStampFName_incr.read(10) == (ENV["START_TIME"][0..9] or "")
	
	lStampFName_incr = Pathname.new("#{lStampFName_incr.to_s}.rotated")
      end
      
      if !lStampFName_incr.exist?
	lStampFName=lStampFName_full
      else
	#get the newest Stamp, either full or incr
	if lStampFName_incr.mtime >= lStampFName_full.mtime
	  lStampFName=lStampFName_incr
	else
	  lStampFName=lStampFName_full
	end
      end
    else
      raise "Unexpected OPT_MODE '#{mode}'"
    end
    
    if lStampFName.exist?
      DateTime.parse(lStampFName.read(19)).strftime("%F %T")
    else
      return nil
    end
  end

  def get_logs
    ret = []
    state_log.each_line {|line|
      fields = line.split(",")
      unless fields.length == 19
	# This is probably because of out of space. ignoring this line.
	$log.info("Ignoring line, too few fields: #{line.inspect}")
	next
      end
            
      hash = {
	:start_time => fields[0],
	:finish_time => fields[1],
	:action => fields[2].downcase.to_sym,
	:set => fields[3],
	:mode => fields[4],
	:pack_method => fields[5],
	:cmd_crypt => fields[6],
	:total_dirs_found => fields[7],
	:total_input_count => fields[8],
	:total_input_size => fields[9],
	:total_output_count => fields[10],
	:total_output_size => fields[11],
	:program_version => fields[12],
	:host => fields[13],
	:domain => fields[14],
	:net => fields[14],
	:crypt_key_id => fields[15],
	:crypt_key_name => fields[16],
	:result => fields[18].strip
      }
      
      if fields[17] =~ /^\".*\"$/
	hash[:target_dir] = Pathname.new(eval(fields[17]))
      else
	hash[:target_dir] = Pathname.new(fields[17])
      end

      if hash[:host] == @host and
	  hash[:domain] == @domain and
	  hash[:set].to_s == @set.to_s then
	ret.push(hash)
      end
    }
    ret
  end

  def get_current_backup_runs(only_completed = false)
    logs = get_logs
    logs = logs.select {|hash|
      $log.debug{"Mode: #{hash[:mode].inspect} Action: #{hash[:action].inspect} => #{hash[:action] == :backup}"}
      hash[:action] == :backup
    }

    latest_full = nil
    logs.each_with_index {|hash, index|
      if hash[:mode] == "full" and hash[:result] != "started" and !(hash[:result] =~ /^safe/)
	$log.debug{"New latest full: #{hash[:start_time]} result: #{hash[:result]}"}
	latest_full = index
      end
    }
   
    if latest_full.nil?
      $log.debug{"No full found."}
      return [] 
    end
    
    logs = logs[latest_full..-1]
    logs = logs.select {|hash| 
      $log.debug{"Backup result: #{hash[:result]}"}
      hash[:result] =~ /^completed/
    } if only_completed
    logs.each_with_index {|log,index|
      raise "Current target directory is already in log!!" if log[:mode] != "full" and @root.to_s == log[:target_dir].to_s
      raise "'#{log[:target_dir]}' does not exist!" unless Pathname.new(log[:target_dir]).exist?
    }
    logs.map {|log| BackupRun.new(log)}
  end
  
  def get_latest_completed_runs    
    return nil unless state_log.exist?
    return @current_completed_runs if @current_completed_runs and (@current_completed_runs_date == state_log.mtime)
    @current_completed_runs_date == state_log.mtime
    @current_completed_runs = get_current_backup_runs(true)
  end

  def latest_full
    runs = get_latest_completed_runs
    runs = runs.select {|r| r[:mode] == "full"}
    runs[-1]
  end
  def latest
    get_latest_completed_runs[-1]    
  end

  def get_latest_dirlist(full_only = nil)
    if full_only
      l = latest_full
    else
      l = latest
    end
    $log.debug{"Latest backup run: #{l.inspect}"}
    return nil if l.nil?

    return nil unless PathList.contains_list?(l[:target_dir],"dirlist")
    ArchivePathList.new(l[:target_dir],"dirlist")
  end

  def get_latest_filelist_for(directory, alternative_root = nil, full_only = nil)
    runs = get_latest_completed_runs
    runs = runs.select {|r| r[:mode] == "full"} if full_only
    runs.reverse.each {|run|
      filelist = run.filelist_for(directory)
      return filelist if filelist
    }
    if alternative_root
      return BackupRun.filelist_for(alternative_root,directory)
    end
    return nil
  end
  
end


EXTRACT_COMMAND = "gpg -q -d '%s' | bunzip2 | cpio --extract --to-stdout --quiet"

class BackupRun

  def BackupRun.filelist_for(root, directory)
    raise "Only relative directories allowed: '#{directory}'" if
      directory.to_s =~ /^\//
    dir = root + directory
    $log.debug{"Looking for file list in #{dir}"}
    return nil unless dir.directory?
    return nil unless PathList.contains_list?(dir)
    $log.debug{"Found file list in #{dir}"}
    ArchivePathList.new(dir)
  end
  def BackupRun.dirlist_for(root)    
    $log.debug{"Looking for dir list in #{root}"}
    
    raise "No pathlist found for backup root #{root}." unless
      PathList.contains_list?(root, "dirlist")
    $log.debug{"Found dir list in #{root}"}
    pl = ArchivePathList.new(root, "dirlist")
  end

  def initialize(log)
    @log = log
    raise "Log for initializing is nil." if log.nil?
    raise "Target directory '#{@log[:target_dir]}' does not exist." unless
      @log[:target_dir].directory?
  end

  def uncrypt(f)
    raise "Could not uncrypt." unless system("gpg","-q","--batch",f)
  end
  def unzip(f)
    raise "Could not unzip." unless system("bunzip2",f)
  end

  def extract(backup_file, filename = nil, target_file = nil)
    $log.debug{"Extracting #{filename or "all files"} from #{backup_file}"}
    raise "No backup file given" unless backup_file
    bf = root + backup_file.to_s
    raise "Backup file #{bf} not found" unless 
      bf.exist?
    
    if filename
      tmp_dir = Pathname.new("/tmp/isirestore-#{Process.pid}")
      FileUtils.mkdir(tmp_dir)
      begin
	FileUtils.cd(tmp_dir) {
	  tmp_file = tmp_dir.to_s + "/restore-#{Process.pid}"
	  FileUtils.copy(bf, tmp_file + ".cpio.bz2.gpg")
	  
	  uncrypt(tmp_file + ".cpio.bz2.gpg")
	  unzip(tmp_file + ".cpio.bz2")
	  
	  execute_command_popen3(["cpio","--quiet","--extract","--preserve-modification-time","--no-absolute-filenames","-F",tmp_file + ".cpio", filename])
	}
	FileUtils.copy(tmp_dir + filename, target_file)
      ensure
	FileUtils.rm_rf(tmp_dir)
      end
      $log.info{"File extracted to '#{target_file}'"}
    else
      fn = "isirestore-#{Process.pid}"
      begin
	FileUtils.copy(bf, "#{fn}.cpio.bz2.gpg")
	uncrypt("#{fn}.cpio.bz2.gpg")
	unzip("#{fn}.cpio.bz2")
	execute_command_popen3(["cpio","--unconditional","--quiet","--extract","--preserve-modification-time","--no-absolute-filenames","-F","#{fn}.cpio"])
      ensure
	Dir.glob("#{fn}.cpio*").each {|f|
	  FileUtils.rm(f)
	}
      end
      $log.info{"File extracted to '#{FileUtils.pwd}'"}
    end

    return target_file
  end

  def extract_directory(directory, target_directory)
    raise "Only relative directories allowed: '#{directory}'" if
      directory.to_s =~ /^\//
    directory = Pathname.new(directory)
    target_directory = Pathname.new(target_directory)
    
    target_directory.mkpath
    $log.info("Extracting #{directory} to #{target_directory}")
    directories_in(directory).each {|d|
      extract_directory(directory + d, target_directory + d)
    }
    
    backup_files = files_in(directory).map {|f| f.backup_file }.uniq.compact
    backup_files.each {|bf|
      FileUtils.cd(target_directory) { 
	begin
	  extract(bf)
	rescue
	  $log.error($!)
	  if $log.debug?
	    $!.backtrace.each {|l| $log.debug(l)}
	  end
	end
      }
    }
  end

  def root; @log[:target_dir]; end
  def log; @log; end

  def filelist_for(directory)
    BackupRun.filelist_for(root,directory)
  end
  
  def dirlist
    BackupRun.dirlist_for(root)
  end
  
  def directories_in(directory)
    raise "Only relative directories allowed: '#{directory}'" if
      directory.to_s =~ /^\//
    dirs = []
    d = root + directory
    d.entries.each {|e|
      if (d + e).directory?
	$log.debug("Found directory #{d + e}")
	next if [".",".."].include?(e.to_s)
	dirs << e
      end
    } if d.directory?
    dirs
  end
  def files_in(directory)
    files = []
    fl = filelist_for(directory)
    (fl or []).each {|e|
      files << e
    }
    files
  end

  def entries_in(directory)    
    raise "Only relative directories allowed: '#{directory}'" if
      directory.to_s =~ /^\//
    files = files_in(directory)
    dirs = directories_in(directory)
    #reg = Regexp.new("^#{Regexp.escape(directory.to_s)}\\/[^\\/]")
    #dirlist.each {|e|
    #ret << e if e.pathname.to_s =~ reg
    #
    [files, dirs]
  end
  
  def [](index)
    @log[index]
  end

end
